README.md
# Clojure Rest MySQL Basic Template

Basic template to run a REST server with Ring Clojure, serving MySQL database, on clojure environment

## Clojure libraries used

- [Ring](https://github.com/ring-clojure/ring)
- [Compojure](https://github.com/weavejester/compojure)
- [Next-JDBC](https://github.com/seancorfield/next-jdbc)
- [Clojure Data](https://github.com/clojure/data.json)

## Installation

```
git clone https://github.com/daryllman/clojure-rest-mysql-basic-template-basic-template
```

## Usage

### 1. Compile to an UberJAR file

```
lein uberjar
```

This will create 2 uberjar files in /target/uberjar

### 2. Run UberJAR file with java

```
java -jar target\uberjar\clojure-rest-mysql-basic-template-0.1.0-SNAPSHOT-standalone.jar [args]
```

### 3. Check if REST Server is working

Go to [http://localhost:8000](). If everything looks good, all is working fine.

### 4. Check MySQL Database is working

See the sample http client calls created under /api-calls/client-api-calls.http  
Just an example:  
[http://localhost:8000/products]()

---

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

### Changed

- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-11-20

### Changed

- Documentation on how to make the widgets.

### Removed

- `make-widget-sync` - we're all async, all the time.

### Fixed

- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-11-20

### Added

- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/clojure-rest-mysql-basic-template-basic-template/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-rest-mysql-basic-template-basic-template/compare/0.1.0...0.1.1


test/clojure_rest_mysql_basic_template/core_test.clj
(ns clojure-rest-mysql-basic-template.core-test
  (:require [clojure.test :refer :all]
            [clojure-rest-mysql-basic-template.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject clojure-rest-mysql-basic-template-basic-template "0.1.0-SNAPSHOT"
  :description "Basic template to run a REST server with Ring Clojure, serving MySQL database, on clojure environment"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [ring "1.8.2"]
                 [compojure "1.6.2"]
                 [org.clojure/java.jdbc "0.7.11"]
                 [seancorfield/next.jdbc "1.1.610"]
                 [mysql/mysql-connector-java "8.0.21"]
                 [org.clojure/data.json "1.0.0"]]
  :main ^:skip-aot clojure-rest-mysql-basic-template.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all
                       :jvm-opts ["-Dclojure.compiler.direct-linking=true"]}})


src/clojure_rest_mysql_basic_template/db/model.clj
(ns clojure-rest-mysql-basic-template.db.model
  (:require [next.jdbc :as jdbc]))

; products Dataset columns
; id - ID of the product
; name - Name of the product
; price - Price of the product
;;; To check data types: https://www.w3schools.com/sql/sql_datatypes.asp

; CRUD OPERATIONS for kindle_products mysql database
; - READ all products
; - READ products of a specified price
; - CREATE a product

(def db {:dbtype "mysql" :dbname "shoe-products"})
(def ds (jdbc/get-datasource db))



; Insert some sample data
(defn load-sample-data [ds]
  (jdbc/execute! ds ["INSERT INTO products (id, name, price)
                      VALUES (?,?,?), (?,?,?), (?,?,?), (?,?,?)"
                     "1001" "Nike Black Shoe" "99"
                     "1002" "Nike White Shoe" "99"
                     "1003" "Puma Running Shoe" "110"
                     "1004" "Puma Walking Shoe" "105"]))

; Create 'products' Table
(defn create-table [ds] ;change to drop table if exists;
  (jdbc/execute! ds ["
                      CREATE TABLE IF NOT EXISTS products (
                      id char(4),
                      name varchar(255),
                      price varchar(5)
                      )"])
  (println "Created products Table")
  (load-sample-data ds)
  (println "Loaded some sample data"))



; Insert
(defn create-product [ds id name price]
  (println "Inserting a row")
  (jdbc/execute!
   ds
   ["INSERT INTO products (id, name, price)
                VALUES (?,?,?)"
    id name price]))



; Read All - wont really be used. Just for testing purpose
(defn read-all-products [ds]
  (println "Reading all products")
  (jdbc/execute! ds ["SELECT * from products"]))


; Read products of a price
(defn read-a-product [ds price]
  (println "Reading products of a price - " price)
  (jdbc/execute! ds ["SELECT * from products WHERE price=?" price]))


src/clojure_rest_mysql_basic_template/db/handler.clj
(ns clojure-rest-mysql-basic-template.db.handler
  (:require [clojure-rest-mysql-basic-template.db.model :refer [read-all-products
                                                                create-product
                                                                read-a-product]]
            [ring.util.response :refer [response]]
            [clojure.data.json :as json]))


; id - ID of the product
; name - Name of the product
; price - Price of the product


(defn handle-all-products [req]
  (let [ds (:clojure-rest-mysql-basic-template/ds req)
        products (read-all-products ds)]
    ;; {:status 200
    ;;  :headers {}
    ;; ;  :body (str products)}))
    ;;  :body (list (json/write-str (response products)))}))
    {:status 200
     :body (list (json/write-str (response products)))}))

(defn handle-a-product [req]
  (let [ds (:clojure-rest-mysql-basic-template/ds req)
        ; id (get-in req [:params "id"])
        ; name (get-in req [:params "name"])
        price (get-in req [:params "price"])
        ; selected-product (read-a-product ds id name price)
        selected-product (read-a-product ds price)]
    {:status 200
     :headers {}
    ;  :body (str products)}))
     :body (list (json/write-str (response selected-product)))}))


(defn handle-create-product [req]
  (let [ds (:clojure-rest-mysql-basic-template/ds req)
        id (get-in req [:params "id"])
        name (get-in req [:params "name"])
        price (get-in req [:params "price"])]
    (create-product ds id name price)
    {:status 200
     :body (str "Successfully inserted a product - id" "(" id "), " "name" "(" name "), " "price" "(" price ")")}))



src/clojure_rest_mysql_basic_template/core.clj
(ns clojure-rest-mysql-basic-template.core
  (:gen-class)
  (:require [clojure-rest-mysql-basic-template.db.model :as products]
            [clojure-rest-mysql-basic-template.db.handler :refer [handle-all-products
                                                                  handle-create-product
                                                                  handle-a-product]])
  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.reload :refer [wrap-reload]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.resource :refer [wrap-resource]]
            [ring.middleware.file-info :refer [wrap-file-info]]
            [compojure.core :refer [defroutes ANY GET POST PUT DELETE]]
            [compojure.route :refer [not-found]]
            [ring.handler.dump :refer [handle-dump]]
            [next.jdbc :as jdbc]))

; Initialise database -> datasource
(def db {:dbtype "mysql" :dbname "shoe-products" :user "root" :password "1234" :serverTimezone "UTC"})
(def ds (jdbc/get-datasource db))


(defn greet [req]
  {:status 200
   :body "Hello world"
   :headers {}})

(defn yo [req]
  (let [name (get-in req [:route-params :name])]
    {:status 200
     :body (str "Yo " name "!")
     :headers {}}))

; REST routes
(defroutes routes
  (GET "/" [] greet)
  (GET "/request" [] handle-dump)
  (GET "/yo/:name" [] yo)

  (GET "/products" [] handle-all-products)
  (POST "/products" [] handle-create-product)
  (POST "/product" [] handle-a-product)

  (not-found "Page not found"))

(defn wrap-db [hdlr]
  (fn [req]
    (hdlr (assoc req :clojure-rest-mysql-basic-template/ds ds))))

(defn wrap-server [hdlr]
  (fn [req]
    (assoc-in (hdlr req) [:headers "Server"] "shoe-products MySQL Server")))

(def app
  (wrap-server (wrap-db (wrap-params routes))))

(defn -main []
  (println "Starting...")
  (products/create-table ds)
  (jetty/run-jetty app {:port (Integer. 8000)}))


;;; For development
; (defn -dev-main []
;   (products/create-table ds)
;   (jetty/run-jetty (wrap-reload #'app) {:port (Integer. 8000)}))


